### ISSUE TITLE
pg: incorrect "delete + join" query since 0.95.8

### ISSUE THREAD

**Issue by Musinux** [knex/knex#4591]
# Environment

Knex version: 0.95.8
Database + version: postgresql 12.7
OS: linux/debian 11

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

I get an incorrect `delete` postgres query, containing the table name in the `delete` statement:
```sql
delete "my_table" from "my_table" left join "users" on "users"."user_id" = "my_table"."user_id" where "my_table"."criteria" < 10
```
It should produce this code instead:
```sql
delete from "my_table" left join "users" on "users"."user_id" = "my_table"."user_id" where "my_table"."criteria" < 10
```
2. Error message
error from postgres:
```
syntax error at or near ""my_table""
```

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.
```javascript
const knex = require('knex')({
	client: 'pg',
	connection: { /* redacted */ }
});

const query = knex
	.select('xyz')
	.from('my_table')
	.leftJoin('users', 'users.user_id', 'my_table.user_id')
        .where('my_table.criteria', '<', 10)
	.delete();

console.log(query.toString());
```

this code produces the following query in 0.95.8:
```sql
delete "my_table" from "my_table" left join "users" on "users"."user_id" = "my_table"."user_id" where "my_table"."criteria" < 10
```

and in 0.95.7:
```sql
delete from "my_table" where "my_table"."criteria" < 10
```
The 0.95.7 result is not correct either, but at least it generates a valid PostgreSQL statement.

I'm not sure the way knex is used in this example is correct, but I found an actual use of knex like this one in my codebase.

I hope this information is relevant, and if not, sorry for the noise.

Best Regards

**Comment by kibertoad** [comment]
This is due to https://github.com/knex/knex/pull/4568
@Egor-Koldasov Could you please take a look?

**Comment by Egor-Koldasov** [comment]
@Musinux  Seems like this query you specified will not work in PG too:
```SQL
delete from "my_table" left join "users" on "users"."user_id" = "my_table"."user_id" where "my_table"."criteria" < 10
```

I tested a similar query it will give an error `syntax error at or near "left"`. I think PostgreSQL does not support such joins in the delete query, can you confirm?

> The 0.95.7 result is not correct either, but at least it generates a valid PostgreSQL statement.

True, although dropping the join silently might result in unintentional data erasure: https://github.com/knex/knex/issues/873#issuecomment-808735961

There is a plan to use `USING` in the case of PG which will result in a query like this
```SQL
delete from "my_table" using "users" where "users"."user_id" = "my_table"."user_id" and "my_table"."criteria" < 10
```
@Musinux  What do you think about this?

**Comment by Musinux** [comment]
@Egor-Koldasov You're totally right, I must have overlooked the expected result too quickly.

I tested the `USING` statement on my own db and it works with a similar query to yours, so I think that would be the right way to go.

However, it might be interesting to alert users when the generated query is likely to be incorrect. I did not know about this incorrect behaviour before yesterday even though this code has been in production :/

Thanks for your time.

**Comment by kibertoad** [comment]
@Egor-Koldasov Will you be able to add USING support, btw? I think @elhigu blessed the approach, so it's just a matter of implementation at this point.

**Comment by Egor-Koldasov** [comment]
@kibertoad I have plans for it, but I might need about a month to get to this issue. If you or anyone wants to implement this it's good.

**Comment by t1a2l** [comment]
@Egor-Koldasov - hey any workaround? can i use using instead ?

**Comment by t1a2l** [comment]
@OlivierCavadenti - it seems that when using "using" i get an error "missing from clause":
mydb(table)
          .select("trip_points.id").first()
          .using("shipments")
          .where("shipments.id", "trip_points.shipment_id")
          .andWhere("shipments.trip_id", "trip_points.trip_id")
          .andWhere("shipments.trip_point_id", "trip_points.id")

**Comment by OlivierCavadenti** [comment]
> @OlivierCavadenti - it seems that when using "using" i get an error "missing from clause": mydb(table) .select("trip_id").first() .using("cars") .where("cars.id", "trip_points.shipment_id") .andWhere("cars.trip_id", "trip_points.trip_id") .andWhere("cars.trip_point_id", "trip_points.id")

See in my PR the syntax, 'using' is for Delete statement. So you need to replace '.select("trip_id").first()' by a 'del().from(...)' syntax.

**Comment by t1a2l** [comment]
@OlivierCavadenti - i am doing delete later, aka: 
var q = mydb(table)
.select("trip_id").first()
.using("cars")
.where("cars.id", "trip_points.shipment_id")
.andWhere("cars.trip_id", "trip_points.trip_id")
.andWhere("cars.trip_point_id", "trip_points.id")
var res = await q;
if(res) q.del();

**Comment by t1a2l** [comment]
@OlivierCavadenti do you understand the isaue?

**Comment by t1a2l** [comment]
Why did you close? Does the PR support my case?

**Comment by OlivierCavadenti** [comment]
It's autoclosing because the PR was merged. Support both classic join syntax and using syntax.

**Comment by t1a2l** [comment]
Thanks. When a new version is going to be published?

**Comment by kibertoad** [comment]
I can make an rc release, but full 1.0.0 still is quite ahead.

**Comment by t1a2l** [comment]
Ok i will wait..

**Comment by kibertoad** [comment]
@OlivierCavadenti Now that delete + join is in, I've been thinking of waiting till better-sqlite dialect PR is finished (hopefully it is, worst case scenario I'll create one myself from a scratch) and also for your improved JSON support, then we do a pass removing deprecated functionality and old Node 10 compatibility code, and we can release 1.0.0.
Do you have any personal preferences of what you would like to see in 1.0.0?

**Comment by OlivierCavadenti** [comment]
@kibertoad After JSON, I will plan 'check' support (not big thing) and maybe could be nice to finish some old/abandonned PRs that just need rebase or tests and can be added.
If we have some times (depending when you see 1.0.0 out) I can see three subjects:
- 'cast' functions support.
- datepart functions.
- lateral join support.

**Comment by kibertoad** [comment]
I would prefer to get it out before the end of the year. Do you think that would be feasible?

**Comment by OlivierCavadenti** [comment]
> I would prefer to get it out before the end of the year. Do you think that would be feasible?

Not sure about all three subjects in the same time, but json + check are sure. For Json I implement select and modification functions but I begin implementations of "where" for json too, and I plan to finish with indexes and json joins. I think it worth the wait for this feature :D.

**Comment by anonrig** [comment]
Hello General Kenobi. I've come with a regression.

![Hello general kenobi](https://media.giphy.com/media/8JTFsZmnTR1Rs1JFVP/giphy.gif)

**Comment by kibertoad** [comment]
Released in 1.0.1.
